### DISCUSSION TITLE
How i can get and manage pEventAttributes from OPC AE and wrapper with Milo?

### DISCUSSION THREAD

**Discussion by Demyanov-A** [eclipse-milo/milo#1253] [General]
Hi!
I am trying to create a project to read alarms and events with milo based on an EventSubscriptionExample.
My test OPC server (matrikon simulator) is a classic COM OPC AE, and I use the OPC AE to UA wrapper to convert data to OPC UA. And everything works fine, events come from the event consumer in the array of Variant, but in addition to the fields I requested, some additional attributes come in the array. How can I identify the data from which attributes I additionally receive? How can I make this data manageable and use it as fields in the event filter?
For example, I can create a query for the required "EventId" field
"new Simple Attribute Operand(
    Identifiers.Base Event Type,
    new QualifiedName[]{new QualifiedName(0, "EventId")},
    Attribute Id.Value.uid(),
    null)".

How can I make the same request for a custom "ExampleAttribute" attribute that exists in the server structure?
Should I create a custom event subtype "ExampleEventType" inherited from the BaseEventType to which this attribute belongs on the server?
Thanks.

**Comment by kevinherron** [comment]
> but in addition to the fields I requested, some additional attributes come in the array. How can I identify the data from which attributes I additionally receive?

Can you explain what you mean by this a little more? Are you saying the `Variant[]` you receive is longer in length than the number of attributes your EventFilter requests? I'm not sure that's even legal.

> How can I make the same request for a custom "ExampleAttribute" attribute that exists in the server structure?
Should I create a custom event subtype "ExampleEventType" inherited from the BaseEventType to which this attribute belongs on the server?

The wrapper is the server, and the server would need to have this type. Is that something you can configure/influence somehow?

**Comment by Demyanov-A** [comment]
> Are you saying the Variant[] you receive is longer in length than the number of attributes your EventFilter requests?

Yes, that's exactly what I meant.
![image](https://github.com/eclipse/milo/assets/64923373/70d6a900-cb60-43b5-9fa3-f6a0e97a9198)
This is what the event looks like in the client. The attributes "OneHundred", "TwoHundred", "ThreeHundred" are displayed as "Attr #0", "Attr #1", "Attr #2".
The eventFilter in my Kotlin code looks like this:

`val eventFields = listOf(
            "EventId",
            "EventType",
            "SourceNode",
            "SourceName",
            "Time",
            "ReceiveTime",
            "Message",
            "Severity",
            "ClientUserId",
            "ConditionName",
            "ConditionClassId",
            "ConditionClassName",
            "Retain",
            "EnabledState",
            "ActiveState",
            "AckedState",
            "Comment"
        )`

    var selectClauses = emptyArray<SimpleAttributeOperand>()
        eventFields.forEach {
            logger.info("Add event field to event filter [${it}]!")
            selectClauses += SimpleAttributeOperand(
                Identifiers.BaseEventType,
                arrayOf(QualifiedName(0, it)),
                AttributeId.Value.uid(),
                null
            )
        }

        eventFilter = EventFilter(selectClauses, ContentFilter(null))

selectClauses.size = 17.
But the output `Variant[]` looks like this:
![image](https://github.com/eclipse/milo/assets/64923373/3c96bf31-657b-4738-a594-7dee0dd595b7)

> The wrapper is the server, and the server would need to have this type.

I know that the wrapper is a server, but I could not find information on how I can get the values of similar attributes in a request using eventFilter. 
I used this documentation - [https://reference.opcfoundation.org/Core/Part9/v105/docs/D.3](url), it says the following:
![image](https://github.com/eclipse/milo/assets/64923373/25694709-b5b9-437e-9ca2-c01ede8a402b)
 In UaExpert client i see that this attributes have COMAEBaseEventType 
![image](https://github.com/eclipse/milo/assets/64923373/5d6cd41d-4606-499f-bfde-db5f402329db)

**Comment by kevinherron** [comment]
I think you'd need the NodeId of the GEAK Event type, then add additional operands to the select clause like:
```
SimpleAttributeOperand(
    GEAK_EVENT_NODE_ID,
    arrayOf(QualifiedName(GEAK_EVENT_NS_IDX, "OneHundred")),
    AttributeId.Value.uid(),
    null
)
```

**Comment by Demyanov-A** [comment]
Thank you, it works!
